### PR TITLE
Refactor must-locksets to use definite mvals instead of addresses

### PR DESCRIPTION
**Pull request by sim642**
Currently must-locksets are (reversed) sets of addresses, which can be the usual addresses (as mvals), but the type also allows `UnknownPtr`, `NullPtr` and `StrPtr`. None of these should be in must-locksets, but currently this is enforced by the combination of three different checks in three different files.

The goal of this PR is to refine the type of must-lockset elements from addresses to mvals with definite indices, whereby this correctness property is enforced by the type system.
**This immediately paid off:** it revealed _two cases_ of unsoundness in recursive mutex handling because definite indices were only checked for the must-lockset, but not for the recursive locking counter map (which must duplicate all the set logic to avoid becoming unsound). See #1432.

Hopefully this also allows some `Addr` wrapping and unwrapping to be avoided, which should be less verbose.

### TODO
- [x] Enforce definite indices in must-locks by type.
- [x] Rename `Simple`.
- [x] Add `all` and `is_all` operations to `Simple`.
- [x] Fix `NULL` warning in cram test.

### CODE REVIEW

**Comment by sim642** [comment]
This goes quite a long way to clean things up, but not as far as I would've liked. The oddity is that privatizations get the mutex as an `AddressDomain` address, which can still be any of the odd things (`UnknownPtr`, `NullPtr` or ambiguous index).

@michael-schwarz What are privatizations supposed to do with those? It seems to me that we completely forgot about the non-definite possibilities, which might make things even unsound. Privatizations may have to be adapted to behave more like the must-lockset analysis: only `lock` definite addresses and `unlock` should unlock all `semantic_equal` addresses in case of non-definite.

**Comment by michael-schwarz** [comment]
> What are privatizations supposed to do with those? It seems to me that we completely forgot about the non-definite possibilities, which might make things even unsound.

I think I considered this issue when designing the privatizations. They should still be sound for the original setting according to the following arguments:

- For `lock` & `mine-W`: Incorporating too many values early makes the analysis imprecise, not unsound; All other operations are based on must-locksets, and for background locksets and minimal locksets only definite elements are kept 
- For `write`: lock is a no-op, other actions are based on must-lockset
- For `per-mutex` flavors incl. relational & protection-based: Protection by one of these indefinite addresses means they are always held when the global is accessed, so that is not harmful either

However, these are indeed a bit shaky in some places and may easily be subject to breakage. So in the long run, it does make sense to refactor this as well.

Also, with the new additions such as `atomic` and invariants about unprotected globals, I'm not sure if this still all holds.

File: src/analyses/mutexAnalysis.ml [hunk 1]
```diff
@@ -146,38 +119,57 @@ struct
...
+            (* TODO: don't repeat the same semantic_equal checks *)
+            (* TODO: rmed per lockset element, not aggregated *)
+            (MustLocksetRW.remove_mval mv s, m')
+          else
+            (s, m')
+        )
         else
-          (s, m')
-      | _ -> (rm s, m)
+          (MustLocksetRW.remove_mval mv s, m) (* TODO: should decrement something if may be recursive? *)
```
Comment (michael-schwarz) [inline src/analyses/mutexAnalysis.ml:162]: No, not really, as nothing was incremented in the first place. Our `MutexTypeAnalysis` is flow-insensitive.
The only thing to consider is when that information changes during iteration. However, since the order is must here, no values where the multiplicity is > 0 for non-must-recursive mutexes should be present at fixpoint.

File: src/cdomain/value/cdomains/mval.ml [hunk 2]
```diff
@@ -36,12 +36,13 @@ struct
   let to_cil_exp lv = Lval (to_cil lv)
 
   let is_definite (_, o) = Offs.is_definite o
-  let top_indices (x, o) = (x, Offs.top_indices o)
 end
 
 module MakeLattice (Offs: Offset.Lattice): Lattice with type idx = Offs.idx =
 struct
   include MakePrintable (Offs)
+
```
Comment (michael-schwarz) [inline src/cdomain/value/cdomains/mval.ml:44]: Why was this moved from Printable?

**Review by michael-schwarz — APPROVED** [review]

File: src/cdomain/value/cdomains/mval.ml (hunk 2 above)
Comment (sim642) [reply src/cdomain/value/cdomains/mval.ml:44]: Because `Printable` indices in offsets aren't required to have a `top`, so `top_indices` cannot be defined for them. That's the case for the new `Z` offsets which are just `Printable`.
Everything else we had before (`IntDomain`, `Unit`, etc) just happened to have `top`.
